src/SLEEF.jl

```diff
@@ -8,6 +8,28 @@ module SLEEF
 
 using Base.Math: uinttype, @horner, exponent_bias, exponent_mask, significand_bits, IEEEFloat, exponent_raw_max
 
+using SIMD
+using SIMD: vifelse
+
+const FloatType64 = Union{Float64,Vec{<:Any,Float64}}
+const FloatType32 = Union{Float32,Vec{<:Any,Float32}}
+const FloatType = Union{FloatType64,FloatType32}
+const IntegerType64 = Union{Int64,Vec{<:Any,Int64}}
+const IntegerType32 = Union{Int32,Vec{<:Any,Int32}}
+const IntegerType = Union{IntegerType64,IntegerType32}
+
+EquivalentInteger(::Type{Float64}) = Int == Int32 ? Int32 : Int64
```

Comment:
I think these should all just return `Int`

can we also rename this to `fpinttype`, the function is quite similar to https://github.com/JuliaLang/julia/blob/master/base/atomics.jl#L331 

Except always returns the machine word size.

If you look at the previous code, we always use `Int`, even for code that operates on `Float32`, because if you use `Int32` for `Float32` inputs of a 64-bit machine, this can crush the range of the calculations for the trig functions (I'm pretty sure, this is true, if I recall correctly).

Reply:
I changed the name from `EquivalentInteger` to `fpinttype`.

If `Int` is `Int64` and you're doing vector operations on 
`Float32`, the resulting integer vectors will be twice the bytes, taking up twice the register space.

This can have a significant impact on runtime. In the case of `exp`, it can add 1-2 nanoseconds.

With 64 bit integers:
```asm
.text
vmovups	(%rsi), %zmm2
movabsq	$139690975831012, %rax  # imm = 0x7F0C56FE23E4
vmulps	(%rax){1to16}, %zmm2, %zmm0
vrndscaleps	$4, %zmm0, %zmm3
vextractf64x4	$1, %zmm3, %ymm0
vcvttps2qq	%ymm0, %zmm0
vcvttps2qq	%ymm3, %zmm1
movabsq	$139690975831052, %rax  # imm = 0x7F0C56FE240C
vbroadcastss	(%rax), %zmm4
movabsq	$139690975831060, %rax  # imm = 0x7F0C56FE2414
vcmpnltps	(%rax){1to16}, %zmm2, %k1
movabsq	$139690975831016, %rax  # imm = 0x7F0C56FE23E8
vcmpnltps	%zmm2, %zmm4, %k2
vfmadd231ps	(%rax){1to16}, %zmm3, %zmm2
movabsq	$139690975831020, %rax  # imm = 0x7F0C56FE23EC
vfmadd231ps	(%rax){1to16}, %zmm3, %zmm2
movabsq	$139690975831024, %rax  # imm = 0x7F0C56FE23F0
vbroadcastss	(%rax), %zmm3
movabsq	$139690975831028, %rax  # imm = 0x7F0C56FE23F4
vfmadd213ps	(%rax){1to16}, %zmm2, %zmm3
movabsq	$139690975831032, %rax  # imm = 0x7F0C56FE23F8
vfmadd213ps	(%rax){1to16}, %zmm2, %zmm3
movabsq	$139690975831036, %rax  # imm = 0x7F0C56FE23FC
vfmadd213ps	(%rax){1to16}, %zmm2, %zmm3
movabsq	$139690975831040, %rax  # imm = 0x7F0C56FE2400
vfmadd213ps	(%rax){1to16}, %zmm2, %zmm3
movabsq	$139690975831044, %rax  # imm = 0x7F0C56FE2404
vfmadd213ps	(%rax){1to16}, %zmm2, %zmm3
vmulps	%zmm2, %zmm2, %zmm4
vmulps	%zmm3, %zmm4, %zmm3
vaddps	%zmm3, %zmm2, %zmm2
movabsq	$139690975831048, %rax  # imm = 0x7F0C56FE2408
vaddps	(%rax){1to16}, %zmm2, %zmm2
vpsraq	$1, %zmm1, %zmm3
vpsraq	$1, %zmm0, %zmm4
movabsq	$139690975831064, %rax  # imm = 0x7F0C56FE2418
vpbroadcastq	(%rax), %zmm5
vpaddq	%zmm5, %zmm4, %zmm6
vpaddq	%zmm5, %zmm3, %zmm7
movabsq	$139690975831104, %rax  # imm = 0x7F0C56FE2440
vmovdqa32	(%rax), %zmm8
vpermt2d	%zmm6, %zmm8, %zmm7
vpslld	$23, %zmm7, %zmm6
vmulps	%zmm6, %zmm2, %zmm2
vpsubq	%zmm3, %zmm1, %zmm1
vpsubq	%zmm4, %zmm0, %zmm0
vpaddq	%zmm5, %zmm0, %zmm0
vpaddq	%zmm5, %zmm1, %zmm1
vpermt2d	%zmm0, %zmm8, %zmm1
vpslld	$23, %zmm1, %zmm0
movabsq	$139690975831056, %rax  # imm = 0x7F0C56FE2410
vbroadcastss	(%rax), %zmm1
vmulps	%zmm0, %zmm2, %zmm1 {%k2}
vmovaps	%zmm1, %zmm0 {%k1} {z}
vmovaps	%zmm0, (%rdi)
movq	%rdi, %rax
vzeroupper
retq
nopw	%cs:(%rax,%rax)
```
This is 60 lines. With 32 bit integers, we have 49 lines:
```asm
.text
vmovups	(%rsi), %zmm0
movabsq	$139690975841852, %rax  # imm = 0x7F0C56FE4E3C
vmulps	(%rax){1to16}, %zmm0, %zmm1
vrndscaleps	$4, %zmm1, %zmm1
vcvttps2dq	%zmm1, %zmm2
movabsq	$139690975841892, %rax  # imm = 0x7F0C56FE4E64
vbroadcastss	(%rax), %zmm3
movabsq	$139690975841900, %rax  # imm = 0x7F0C56FE4E6C
vcmpnltps	(%rax){1to16}, %zmm0, %k1
movabsq	$139690975841856, %rax  # imm = 0x7F0C56FE4E40
vcmpnltps	%zmm0, %zmm3, %k2
vfmadd231ps	(%rax){1to16}, %zmm1, %zmm0
movabsq	$139690975841860, %rax  # imm = 0x7F0C56FE4E44
vfmadd231ps	(%rax){1to16}, %zmm1, %zmm0
movabsq	$139690975841864, %rax  # imm = 0x7F0C56FE4E48
vbroadcastss	(%rax), %zmm1
movabsq	$139690975841868, %rax  # imm = 0x7F0C56FE4E4C
vfmadd213ps	(%rax){1to16}, %zmm0, %zmm1
movabsq	$139690975841872, %rax  # imm = 0x7F0C56FE4E50
vfmadd213ps	(%rax){1to16}, %zmm0, %zmm1
movabsq	$139690975841876, %rax  # imm = 0x7F0C56FE4E54
vfmadd213ps	(%rax){1to16}, %zmm0, %zmm1
movabsq	$139690975841880, %rax  # imm = 0x7F0C56FE4E58
vfmadd213ps	(%rax){1to16}, %zmm0, %zmm1
movabsq	$139690975841884, %rax  # imm = 0x7F0C56FE4E5C
vfmadd213ps	(%rax){1to16}, %zmm0, %zmm1
vmulps	%zmm0, %zmm0, %zmm3
vmulps	%zmm1, %zmm3, %zmm1
vaddps	%zmm1, %zmm0, %zmm0
movabsq	$139690975841888, %rax  # imm = 0x7F0C56FE4E60
vaddps	(%rax){1to16}, %zmm0, %zmm0
vpsrld	$1, %zmm2, %zmm1
vpslld	$23, %zmm1, %zmm3
vpbroadcastd	(%rax), %zmm4
vpaddd	%zmm4, %zmm3, %zmm3
vmulps	%zmm3, %zmm0, %zmm0
vpsubd	%zmm1, %zmm2, %zmm1
vpslld	$23, %zmm1, %zmm1
vpaddd	%zmm4, %zmm1, %zmm1
movabsq	$139690975841896, %rax  # imm = 0x7F0C56FE4E68
vbroadcastss	(%rax), %zmm2
vmulps	%zmm1, %zmm0, %zmm2 {%k2}
vmovaps	%zmm2, %zmm0 {%k1} {z}
vmovaps	%zmm0, (%rdi)
movq	%rdi, %rax
vzeroupper
retq
nopl	(%rax)
```

Here are the parts related to this, in the 64 bit int version:
```asm
vrndscaleps	$4, %zmm0, %zmm3
vextractf64x4	$1, %zmm3, %ymm0
vcvttps2qq	%ymm0, %zmm0
vcvttps2qq	%ymm3, %zmm1
...
vpsraq	$1, %zmm1, %zmm3
vpsraq	$1, %zmm0, %zmm4
movabsq	$139690975831064, %rax  # imm = 0x7F0C56FE2418
vpbroadcastq	(%rax), %zmm5
vpaddq	%zmm5, %zmm4, %zmm6
vpaddq	%zmm5, %zmm3, %zmm7
movabsq	$139690975831104, %rax  # imm = 0x7F0C56FE2440
vmovdqa32	(%rax), %zmm8
vpermt2d	%zmm6, %zmm8, %zmm7
vpslld	$23, %zmm7, %zmm6
vmulps	%zmm6, %zmm2, %zmm2
vpsubq	%zmm3, %zmm1, %zmm1
vpsubq	%zmm4, %zmm0, %zmm0
vpaddq	%zmm5, %zmm0, %zmm0
vpaddq	%zmm5, %zmm1, %zmm1
vpermt2d	%zmm0, %zmm8, %zmm1
vpslld	$23, %zmm1, %zmm0
```
32 bit int:
```asm
vrndscaleps	$4, %zmm1, %zmm1
vcvttps2dq	%zmm1, %zmm2
...
vpsrld	$1, %zmm2, %zmm1
vpslld	$23, %zmm1, %zmm3
vpbroadcastd	(%rax), %zmm4
vpaddd	%zmm4, %zmm3, %zmm3
vmulps	%zmm3, %zmm0, %zmm0
vpsubd	%zmm1, %zmm2, %zmm1
vpslld	$23, %zmm1, %zmm1
```
Allocating and operating on two registers instead of 1.


However, you're are correct:
```julia
julia> using SLEEF, SIMD

julia> x = Vec{16,Float32}((1f3,1f5,1f7,1f9,1f11,1f13,1f15,1f17,1f19,1f21,1f23,1f25,1f27,1f29,1f31,1f33))
<16 x Float32>[1000.0, 100000.0, 1.0e7, 1.0e9, 1.0e11, 1.0e13, 1.0e15, 1.0e17, 1.0e19, 1.0e21, 1.0e23, 1.0e25, 1.0e27, 1.0e29, 1.0e31, 1.0e33]

julia> SLEEF.sin(x)
<16 x Float32>[0.82687956, 0.0357488, 0.13669702, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0]

julia> using SLEEFwrap

julia> SLEEFwrap.sin(x.elts)
(VecElement{Float32}(0.82687956f0), VecElement{Float32}(0.0357488f0), VecElement{Float32}(0.42054778f0), VecElement{Float32}(0.5458434f0), VecElement{Float32}(0.99810874f0), VecElement{Float32}(0.96887577f0), VecElement{Float32}(0.9944343f0), VecElement{Float32}(-0.5699717f0), VecElement{Float32}(0.5780979f0), VecElement{Float32}(0.7704365f0), VecElement{Float32}(-0.925232f0), VecElement{Float32}(-0.40585858f0), VecElement{Float32}(-0.97865087f0), VecElement{Float32}(0.8592228f0), VecElement{Float32}(-0.039693512f0), VecElement{Float32}(0.33392745f0))

julia> using BenchmarkTools

julia> @benchmark SLEEFwrap.sin($x.elts)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  4400699998796492385
  --------------
  minimum time:     77.960 ns (0.00% GC)
  median time:      79.492 ns (0.00% GC)
  mean time:        79.636 ns (0.00% GC)
  maximum time:     125.621 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     968

julia> @benchmark SLEEF.sin($x).elts
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  4400699998796492385
  --------------
  minimum time:     19.888 ns (0.00% GC)
  median time:      19.992 ns (0.00% GC)
  mean time:        20.038 ns (0.00% GC)
  maximum time:     49.129 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     997
```
SLEEF (the C library)'s `sin` still has solid range on these trig functions, and gets dramatically better performance anyway.

Perhaps we use 32 bit integers for functions like `exp`, and 64 bit integers for the periodic trig functions?

Reply:
I replaced all instances of `fpinttype(T)` with `Int` in the trig file (locally), but:
```julia
julia> SLEEF.sin(x)
<16 x Float32>[0.82687956, 0.0357488, 0.13669702, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0]
```
is still the answer I get.
That is because of line 213:
```julia
u = vifelse((!isinf(t)) & (isnegzero(t) | (abs(t) > TRIG_MAX(T))), T(-0.0), u)
```
and `SLEEF.TRIG_MAX(Float32)` returning `1.0f7`.
However, that still does not explain:
```julia
julia> (SLEEF.sin(1e7),SLEEF.sin(1f7),SLEEF.sin_fast(1f7),sin(1e7))
(0.4205477931907825, 0.13669702f0, 0.4205478f0, 0.4205477931907825)
```
why `SLEEF.sin` is getting the wrong answer here.

Checking out the latest master (rather than this PR)...
```julia
julia> (SLEEF.sin(1e7),SLEEF.sin(1f7),SLEEF.sin_fast(1f7),sin(1e7))
(0.4205477931907825, 0.13669702f0, 0.4205478f0, 0.4205477931907825)
```
so it is an existing problem.

If you're using `TRIG_MAX(::Type{Float32}) = 1f7`, then 32 bit integers should be okay.

Reply:
Ok I see, but the trig function is only guaranteed over 
 

> Notes
The trigonometric functions are tested to return values with specified accuracy when the argument is within the following range:

> Double (Float64) precision trigonometric functions : [-1e+14, 1e+14]
> Single (Float32) precision trigonometric functions : [-39000, 39000]

not `1f7`



If I recall correctly it's a lot faster for non-vectorized code to _always_ use  machine size-int, even if operating on 32 bit floats.

Based on your analysis this is not true for the vector versions?



However, according to your analysis this is not true for vector versions.

Reply:
AFAIK, 32 and 64 bit operations should be the same fast when not vectorized -- as long as you don't need to promote from one to the other. (Note that pointers on 64 bit machines are 64 bits.)

The difference is when they are vectorized. Half the bits means you can fit twice as many into a register, and operate on twice as many per operation.